torch/__init__.py

```diff
@@ -2587,6 +2589,43 @@ def get_device_module(device: _Optional[_Union[torch.device, str]] = None):
     return device_module
 
 
+class DeviceGuard:
+    def __init__(self, device_index: builtins.int):
+        self.idx = device_index
+        self.prev_idx = -1
+
+    def __enter__(self):
+        self.prev_idx = torch._C._exchange_device(self.idx)
+
+    def __exit__(self, type: _Any, value: _Any, traceback: _Any):
+        self.idx = torch._C._maybe_exchange_device(self.prev_idx)
+        return False
+
+
+class StreamGuard:
+    def __init__(self, stream: torch.Stream):
+        self.stream = stream
+        self.src_prev_stream = None
+        self.dst_prev_stream = None
+
+    def __enter__(self):
+        self.src_prev_stream = torch.current_stream()  # type: ignore[assignment]
+
+        # If the stream is not on the current device, then
+        # set the current stream on the device
+        if self.src_prev_stream.device != self.stream.device:  # type: ignore[attr-defined]
+            with DeviceGuard(self.stream.device.index):
+                self.dst_prev_stream = torch.current_stream()  # type: ignore[assignment]
+        torch.set_stream(self.stream)
+
+    def __exit__(self, type: _Any, value: _Any, traceback: _Any):
+        # Reset the stream on the original device and destination device
+        if self.src_prev_stream.device != self.stream.device:  # type: ignore[attr-defined]
+            torch.set_stream(self.dst_prev_stream)  # type: ignore[arg-type]
+        torch.set_stream(self.src_prev_stream)  # type: ignore[arg-type]
+        return False
```

Comment:
This seems a bit weird to me, maybe adding a new internal API called _exchange_stream would be a better choice?